### DISCUSSION TITLE
Bug related to using js functions and x-text

### DISCUSSION THREAD

**Discussion by kacperwyczawski** [alpinejs/alpine#3915] [7. Bugs (archived)] [answered]
# Code

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Alpine.js</title>

    <script>
      function range(a, b) {
        let result = '';
        for (let i = a; i <= b; i++) {
          result += `${i} `;
        }
        return result;
      }
    </script>
    <script
      defer
      src="https://cdn.jsdelivr.net/npm/alpinejs@3.13.3/dist/cdn.min.js"
    ></script>
  </head>
  <body>
    <section x-data="{a: 0, b: 0}">
      a: <input type="number" x-model="a" placeholder="0" /> b:
      <input type="number" x-model="b" placeholder="0" />
      <br />
      dynamic a:
      <pre x-text="range(a, b)"></pre>
      static a:
      <pre x-text="range(4, b)"></pre>
    </section>
  </body>
</html>

```

# Steps to reproduce

1. Run this code
2. Type `4` into the first input (a)
3. Type `9` into the second input (b)
4. For now everything should work as expected
5. Increment value of second input by 1
6. As you can see, `<pre>` labelled *dynamic a* no longer displays range

# Browsers

- Firefox 120.0.1
- Firefox for Android 121.0b7

# Stackblitz reproduction

https://stackblitz.com/edit/web-platform-manu3a?file=index.html

**Comment by hans-lts** [comment]
## Explanation

So this ends up not being an Alpine issue.

According to the [spec](https://developer.mozilla.org/en-US/docs/Web/API/InputEvent/data), `<input>` fields always return a `string` as the value's datatype. That means when `range(a, b)` is invoked, it compares two _strings_ instead of two _integers_ which causes weird results.

#### Expectation (two integers)
`range(4, 10)` 🎉

#### Reality (two strings)
`range('4', '10')` ❌

If you want to know more about how javascript handles string comparisons check it out here: https://javascript.info/comparison#string-comparison.

## Solution

Use `parseInt(x)` inside the `range(a, b)` function to transform the input to a numerical value.

```js
function range(a, b) {
        let result = '';

        let int_a = parseInt(a);
        let int_b = parseInt(b);

        if (int_b < int_a) {
          return 'invalid range';
        }

        for (let i = int_a; i <= int_b; i++) {
          result += `${i} `;
        }

        return result;
}
```

### Example

https://stackblitz.com/edit/web-platform-or3upm?file=index.html

**Reply by ekwoka** [reply, marked as answer]
Easier and more correct to just use `x-model.number`

**Reply by kacperwyczawski** [reply]
Thanks and sorry for unnecessary issue

**Comment by ekwoka** [comment]
Just a normal javascript thing.

you're comparing strings, not numbers.

Change your x-model to `x-model.number` and it will work.

https://alpinejs.dev/directives/model#number
